esp-hal-common/src/soc/esp32s3/mod.rs

```diff
@@ -29,3 +33,89 @@ pub(crate) mod constants {
     pub const SOC_DRAM_LOW: u32 = 0x3FC8_8000;
     pub const SOC_DRAM_HIGH: u32 = 0x3FD0_0000;
 }
+
+#[cfg(feature = "rt")]
+#[doc(hidden)]
+#[link_section = ".rwtext"]
+pub unsafe fn configure_cpu_caches() {
+    // this is just the bare minimum we need to run code from flash
+    // consider implementing more advanced configurations
+    // see https://github.com/apache/incubator-nuttx/blob/master/arch/xtensa/src/esp32s3/esp32s3_start.c
+
+    extern "C" {
+        fn rom_config_instruction_cache_mode(
+            cfg_cache_size: u32,
+            cfg_cache_ways: u8,
+            cfg_cache_line_size: u8,
+        );
+    }
+
+    // ideally these should be configurable
+    const CONFIG_ESP32S3_INSTRUCTION_CACHE_SIZE: u32 = 0x4000; // ESP32S3_INSTRUCTION_CACHE_16KB
+    const CONFIG_ESP32S3_ICACHE_ASSOCIATED_WAYS: u8 = 8; // ESP32S3_INSTRUCTION_CACHE_8WAYS
+    const CONFIG_ESP32S3_INSTRUCTION_CACHE_LINE_SIZE: u8 = 32; // ESP32S3_INSTRUCTION_CACHE_LINE_32B
+
+    // Configure the mode of instruction cache: cache size, cache line size.
+    rom_config_instruction_cache_mode(
+        CONFIG_ESP32S3_INSTRUCTION_CACHE_SIZE,
+        CONFIG_ESP32S3_ICACHE_ASSOCIATED_WAYS,
+        CONFIG_ESP32S3_INSTRUCTION_CACHE_LINE_SIZE,
+    );
+}
+
+/// Function initializes ESP32S3 specific memories (RTC slow and fast) and
+/// then calls original Reset function
+///
+/// ENTRY point is defined in memory.x
+/// *Note: the pre_init function is called in the original reset handler
+/// after the initializations done in this function*
+#[cfg(feature = "rt")]
```

Comment:
there is no such feature in esp-hal-common so we don't get ESP32Reset function - makes the code crash immediately

(also in other places)